### PR TITLE
🌱 Check for build metadata in cert-manager version 

### PR DESCRIPTION
**Pull request by g-gaston**
**What this PR does / why we need it**:
As an operator, I would like to upgrade `cert-manager` using `clusterctl` even when the version hasn't changed. For example, if I created a new build due to a change in a downstream dependency (imagine security updates in a base image) but I'm still using the same upstream code.

In order to work around this, I could artificially bump the version, but that, aside from being confusing, could be problematic if in the future I need to upgrade to the real version.

The pre-release section could also be used for to work around this, but I find it confusing as well since these are not necessarily pre-release builds, they are production ready.

I propose to use the build metadata in the semver (which fits very well this usecase) to detect such changes. This PR just adds an extra check when the semvers have the same precedence, and if the build metadata is different, trigger an upgrade.

### CODE REVIEW

File: cmd/clusterctl/client/cluster/cert_manager.go
Comment (fabriziopandini) [inline cmd/clusterctl/client/cluster/cert_manager.go:331]: Have you considered to use `util.Compare(a, b, WithBuildTags())` instead?
@ykakarap opinions ^^

Comment (g-gaston) [reply cmd/clusterctl/client/cluster/cert_manager.go:331]: I just tried that. The problem is that method uses string comparison to determine precedence. That works if you use more normalized build metadata schemes like `v0.0.1+build.1` -> `v0.0.1+build.2`. But it won't work with others, like using a commit hash `v0.0.1+hf87h89` -> `v0.0.1+hf87h88`

I was aiming for something more simple that supports the latter

Comment (ykakarap) [reply cmd/clusterctl/client/cluster/cert_manager.go:331]: @g-gaston The only case where this PR and using `util.Compare(a, b, WithBuildTags())` will defer will be if the new build tag is ordered higher in string compare but you still want `needUpgrade` to be set to `true` (because it changed).

If you want to achieve the above case then current approach is the way to go 👍🏼

Comment (g-gaston) [reply cmd/clusterctl/client/cluster/cert_manager.go:331]: @ykakarap yeah, I couldn't have said better. Thanks!

Comment (fabriziopandini) [reply cmd/clusterctl/client/cluster/cert_manager.go:331]: What about adding an WithUnsortedBuildTags option?
I really would like to keep version compare logic in a single place given that we just completed a cleanup effort for making this consistent across the codebase

Comment (vincepri) [reply cmd/clusterctl/client/cluster/cert_manager.go:331]: Should build metadata even be considered sortable? By default build information isn't considered sortable, what would be the downside to only catch difference in build metadata rather than having an ordered comparison?

Comment (g-gaston) [reply cmd/clusterctl/client/cluster/cert_manager.go:331]: @fabriziopandini 
I can add that option. What would be the behavior: if normal `Compare` returns 0 and build tags differ, always return -1?

However, I think I agree with @vincepri and I find a bit confusing taking build metadata into account for an ordered comparison.

If the goal is to keep all the semver logic in the `version` package, maybe the solution comes by adding a new method like `version.IsNewerOrSameAndDifferentMetadata`? (pls advice for a better name if you an idea). At the end, we don't need an ordered comparison in this scenario, we just need a true/false for this specific condition.

Comment (fabriziopandini) [reply cmd/clusterctl/client/cluster/cert_manager.go:331]: Sorted metadata is a requirement from folks using versioning schema like v0.0.1+build.1 -> v0.0.1+build.2, which I kind of agree it is a stretch, but we can't drop this without breaking users expectations.

I prefer a single compare method with many explicit options to having different methods; however, I'm happy to consider better options names if this helps to make things more clear

Comment (vincepri) [reply cmd/clusterctl/client/cluster/cert_manager.go:331]: Makes sense, thanks!

Comment (vincepri) [reply cmd/clusterctl/client/cluster/cert_manager.go:331]: One note though is that now we're behaving in different ways in different places, is there a way that we could do a smarter detection if strings shouldn't be compared with `>`, but if should just be different?

Comment (fabriziopandini) [reply cmd/clusterctl/client/cluster/cert_manager.go:331]: not sure.
TBH now that I think about this different use cases, there is a possible conflict between this changeset and requirement from folks using versioning schema like v0.0.1+build.1 -> v0.0.1+build.2.

If they apply the same versioning schema to cert manager (and I'm pretty sure there are users doing this), they will expect also cert manager upgrade to consider sorted metadata to decide when to upgrade - a thing that is not implemented now, but it seems a sort of bug now that I think about it -

This PR instead proposes to just check for metadata equality, without sorting, but this will give unexpected results for people using v0.0.1+build.1 naming with cert manager.

My first reaction to this realisation is that we should use sorted metadata also here, so at least we are consistent, but as downside, we should somehow document that tagging custom-built components with hashes is not supported...

@vincepri @ykakarap opinions?

Comment (sbueringer) [reply cmd/clusterctl/client/cluster/cert_manager.go:331]: I might be misinterpreting this but here's a quote from sem ver 2.0.0:
> Build metadata MUST be ignored when determining version precedence. Thus two versions that differ only in the build metadata, have the same precedence. Examples: 1.0.0-alpha+001, 1.0.0+20130313144700, 1.0.0-beta+exp.sha.5114f85, 1.0.0+21AF26D3—-117B344092BD.
> [...]
> Precedence MUST be calculated by separating the version into major, minor, patch and pre-release identifiers in that order (Build metadata does not figure into precedence).
> https://semver.org/

Comment (fabriziopandini) [reply cmd/clusterctl/client/cluster/cert_manager.go:331]: > Sounds like we violate the semver spec when we support ordering by build metadata.

Kind of, but this was accepted given the requirements from folks using versioning schema like v0.0.1+build.1 -> v0.0.1+build.2 and we are already supporting this all across the codebase.
Given that IMO it is a bug we are not supporting the extended version precedence rule also here.

Comment (sbueringer) [reply cmd/clusterctl/client/cluster/cert_manager.go:331]: Agree. We shouldn't be inconsistent here when we intentionally did it in another parts of the code base. I wasn't aware of that.

Comment (g-gaston) [reply cmd/clusterctl/client/cluster/cert_manager.go:331]: Ok, got it. With all this new info/requirements, I have a different proposal:
* Use `util.Compare(a, b, WithBuildTags())` for cert-manager `shouldUpgrade`
* When comparing `buildIdentifier`s in the `compare` method, if both are strings:
  * If equal, return 0
  * If not equal, return -1

With the proposed solution:
* We still support `v0.0.1+build.1` -> `v0.0.1+build.2` usecase (it will behave the same way it does now)
* We support hashes in the build metadata
* We keep all the semver compare logic in one place (the `version` package)
* The only thing that changes is if folks try to do something like `v0.0.1+build2` -> `v0.0.1+build1`. That would return -1 with the proposed solution, when right now returns 1. But that doesn't seem like a usecase that needs to be supported

If it seems to risky to change the behavior of `WithBuildTags()`, we can create a different option

**Comment by enxebre** [comment]
slightly related https://github.com/kubernetes-sigs/cluster-api/issues/5490
/milestone v1.1

File: cmd/clusterctl/client/cluster/cert_manager.go
Comment (ykakarap) [reply cmd/clusterctl/client/cluster/cert_manager.go:331]: How about supporting comparison strategies in `WithBuildTags()`? So the option `WithBuildTags()` becomes `WithBuildTags(strategy string)` the list of supported strategies are `AlphaNumeric` and `Random`. 
- `AlphaNumeric` should be the default strategy (the strategy we currently use). 
- `Random` will return -1 as long as the build tags are different.

Comment (g-gaston) [reply cmd/clusterctl/client/cluster/cert_manager.go:331]: That sounds reasonable. However, I were to use the `Random` strategy, cert-manager wouldn't support the `v0.0.1+build.1` -> `v0.0.1+build.2` usecase, and it seemed like @fabriziopandini was very inclined to support it here as well. In fact, it would support it, but it would also support `v0.0.1+build.2` -> `v0.0.1+build.1`, which I believe is unwanted.

Comment (ykakarap) [reply cmd/clusterctl/client/cluster/cert_manager.go:331]: You are right, I don't believe we should support different strategies. From @fabriziopandini comments it looks like we want to be consistent across our code base in how we compare versions. Within our codebase we currently use  build tags as well to compare versions (`util.Compare(a, b, WithBuildTags())`) and we should continue doing it.

I am +1 with @fabriziopandini's comment that we should be consistent in the version compare method we use everywhere and document that custom-built components with hashes is not supported.
>My first reaction to this realisation is that we should use sorted metadata also here, so at least we are consistent, but as downside, we should somehow document that tagging custom-built components with hashes is not supported...


This would mean cert-manger will also use the existing `WithBuildTags()` way of doing things.

Comment (g-gaston) [reply cmd/clusterctl/client/cluster/cert_manager.go:331]: Just for my own understanding, what's the reason to perform string ordered comparison right now?
I totally understand the `v0.0.1+build.1` -> `v0.0.1+build.2` usecase, but that doesn't require it, it only requires ordered comparison for the numbers
I'm wondering if there is an actual usecase to support that implementation or if it is something that can potentially be changed elsewhere, achieving both consistency across the codebase and hash support

Comment (g-gaston) [reply cmd/clusterctl/client/cluster/cert_manager.go:331]: @ykakarap yeah, exactly!
I'll go ahead and implement the changes I'm proposing and we can do another round of discussions over them
It'll probably be easier that way :)

Comment (g-gaston) [reply cmd/clusterctl/client/cluster/cert_manager.go]: @ykakarap done!

File: cmd/clusterctl/client/cluster/cert_manager.go
Comment (vincepri) [inline cmd/clusterctl/client/cluster/cert_manager.go:304]: Can we define `needUpgrade` and `curentVersion` variable on line 310?

File: cmd/clusterctl/client/cluster/cert_manager.go
Comment (vincepri) [inline cmd/clusterctl/client/cluster/cert_manager.go:351]: Could we use only semver.Version everywhere and avoid carrying strings as well?

File: cmd/clusterctl/client/cluster/cert_manager.go
Comment (vincepri) [inline cmd/clusterctl/client/cluster/cert_manager.go:306]:
```suggestion
	desiredVersionStr := config.Version()
	desiredVersion, err := semver.ParseTolerant(desiredVersionStr)
```

File: util/version/version.go
Comment (vincepri) [inline util/version/version.go:168]: Add a comment here to explain this logic and why we're never returning `1`

File: cmd/clusterctl/client/cluster/cert_manager.go
Comment (g-gaston) [reply cmd/clusterctl/client/cluster/cert_manager.go]: `semver.Version` doesn't have a field with the unparsed semver
I could use the `String()` method to regenerate it whenever it's needed, but that seems a bit unnecessary, considering we already have it

File: cmd/clusterctl/client/cluster/cert_manager_test.go
Comment (ykakarap) [inline cmd/clusterctl/client/cluster/cert_manager_test.go:273]: Can you please also add a test case where the build metadata is non-numeric but equal?
Would be nice to call out that case.

**Comment by g-gaston** [comment]
@ykakarap @vincepri 
I think I addressed your comments. Anything else I can do to help move this forward? Thanks! :)

**Comment by ykakarap** [comment]
@g-gaston The changes look good.
Looks like you have 2 different commits because they do different things but I think the practice is to squash into 1 commit.

LGTM pending squash.

**Comment by sbueringer** [comment]
> @g-gaston The changes look good. Looks like you have 2 different commits because they do different things but I think the practice is to squash into 1 commit.
> 
> LGTM pending squash.

I think it's okay to keep multiple commits if there is a reason to split up the implementation that way. It's just that usually there isn't :)

**Comment by ykakarap** [comment]
>I think it's okay to keep multiple commits if there is a reason to split up the implementation that way. It's just that usually there isn't :)

Good to know. :) 

/lgtm

**Review by ykakarap — APPROVED** [review]
/lgtm

**Comment by g-gaston** [comment]
@ykakarap @sbueringer 
I think there is a workflow that needs approval in order to be run? Thanks! 🙏🏻

**Comment by ykakarap** [comment]
@fabriziopandini PTAL.

**Comment by g-gaston** [comment]
> @ykakarap @sbueringer I think there is a workflow that needs approval in order to be run? Thanks! 🙏🏻

> @fabriziopandini PTAL.

@fabriziopandini 🙏🏻

**Comment by fabriziopandini** [comment]
@g-gaston sorry but this PR somehow went out of the radar. I will try to provide a feedback by EOW

File: util/version/version.go
Comment (fabriziopandini) [inline util/version/version.go:165]: In order to surface this logic could you kindly update the comment for the `WithBuildTags` with something similar to:
```
// WithBuildTags modifies the version comparison to also consider build tags
// when comparing versions.
// Performs a standard version compare between a and b. If the versions
// are equal, build identifiers will be used to compare further; precedence for two build
// identifiers is determined by comparing each dot-separated identifier from left to right
// until a difference is found as follows:
// - Identifiers consisting of only digits are compared numerically.
// - Numeric identifiers always have lower precedence than non-numeric identifiers.
// - Identifiers with letters or hyphens are compared only for equality, otherwise, 2 is returned given
//.  that it is not possible to identify if lower or greater (non-numeric identifiers could be random build
//.  identifiers).
```

Other comments should be modified accordingly by adding
```
// 2 == v is different than o (it is not possible to identify if lower or greater).
```

File: util/version/version.go
Comment (fabriziopandini) [inline util/version/version.go:171]: What about returning 2 in this case, indicated that things are different (given that -1 indicates that  -1 == v is less than o, which is not correct)

Comment (fabriziopandini) [reply util/version/version.go:171]: Note: this requires a change in the `isUpgrading` func in`controllers/topology/contract/controlplane.go`, from 

```
return version.Compare(specV, statusV, version.WithBuildTags()) == 1, nil
```

to

```
// NOTE: we are considering the control plane upgrading when the version is greater
// or when the version has a different build metadata.
return version.Compare(specV, statusV, version.WithBuildTags()) >= 1, nil
```

**Review by fabriziopandini — COMMENTED** [review]
@g-gaston I'm ok with the proposed solution with a small improvement that makes it explicit and distinct from other values when we consider things different (without a notion of order)

Also, please squash commits

File: util/version/version.go
Comment (g-gaston) [reply util/version/version.go]: Done

File: util/version/version.go
Comment (g-gaston) [reply util/version/version.go]: I like it, done

**Comment by g-gaston** [comment]
@fabriziopandini Feedback comments addressed. Also, squashed commits
PTAL 🙂

**Comment by fabriziopandini** [comment]
/lgtm
@sbueringer @ykakarap for a final pass

**Comment by sbueringer** [comment]
/lgtm

@g-gaston Thx for the patience!!

File: util/version/version.go
Comment (ykakarap) [inline util/version/version.go:209]: Nice work on a well detailed comment on the function 👍🏼

**Review by ykakarap — APPROVED** [review]
/lgtm

Great work! Thank you @g-gaston

**Pull request merged**
